File: ios/FLEXR/Data/Remote/FlexrAPI.swift
import Foundation

/// Vollständige Abbildung der FastAPI-Router aus backend/app/routers/.
/// Ein Aufruf pro Endpunkt, gleiche Pfade wie im Web-Frontend und in der
/// Android-App.
struct FlexrAPI {

    private let client: APIClient

    init(client: APIClient) {
        self.client = client
    }

    // MARK: - auth.py

    func register(_ body: RegisterRequestDTO) async throws -> TokenResponseDTO {
        try await client.send(.post, "api/auth/register", body: body)
    }

    func login(_ body: LoginRequestDTO) async throws -> TokenResponseDTO {
        try await client.send(.post, "api/auth/login", body: body)
    }

    /// Macht eine Selbstlöschung innerhalb der 30-Tage-Karenzzeit rückgängig.
    /// Nimmt dieselben Zugangsdaten wie [login] entgegen (siehe
    /// routers/auth.reactivate).
    func reactivate(_ body: LoginRequestDTO) async throws -> TokenResponseDTO {
        try await client.send(.post, "api/auth/reactivate", body: body)
    }

    /// Zurücksetz-Link per Mail anfordern. Das Zurücksetzen selbst passiert im Browser.
    func forgotPassword(_ body: PasswordForgotRequestDTO) async throws -> OkResponseDTO {
        try await client.send(.post, "api/auth/password/forgot", body: body)
    }

    // MARK: - profiles.py

    /// Beendet alle anderen Sitzungen - die Antwort trägt einen frischen Token.
    func changePassword(_ body: PasswordChangeRequestDTO) async throws -> TokenResponseDTO {
        try await client.send(.post, "api/profiles/me/password", body: body)
    }

    func changeEmail(_ body: EmailChangeRequestDTO) async throws -> MyProfileDTO {
        try await client.send(.post, "api/profiles/me/email", body: body)
    }

    func myProfile() async throws -> MyProfileDTO {
        try await client.send(.get, "api/profiles/me")
    }

    func updateMyProfile(_ body: UpdateProfileRequestDTO) async throws -> MyProfileDTO {
        try await client.send(.patch, "api/profiles/me", body: body)
    }

    /// DELETE mit Körper — vom Backend so vorgesehen (Passwortbestätigung).
    func deleteMyAccount(_ body: DeleteAccountRequestDTO) async throws {
        try await client.send(.delete, "api/profiles/me", body: body)
    }

    func myConsents() async throws -> [ConsentDTO] {
        try await client.send(.get, "api/profiles/me/consents")
    }

    func revokeConsent(_ body: ConsentRevokeRequestDTO) async throws -> ConsentRevokeResponseDTO {
        try await client.send(.post, "api/profiles/me/consents/revoke", body: body)
    }

    func grantConsent(_ body: ConsentGrantRequestDTO) async throws -> ConsentGrantResponseDTO {
        try await client.send(.post, "api/profiles/me/consents/grant", body: body)
    }

    func presignPhoto(_ body: PresignPhotoRequestDTO) async throws -> PresignPhotoResponseDTO {
        try await client.send(.post, "api/profiles/me/photos/presign", body: body)
    }

    func addPhoto(_ body: AddPhotoRequestDTO) async throws -> MyProfileDTO {
        try await client.send(.post, "api/profiles/me/photos", body: body)
    }

    func deletePhoto(id: String) async throws -> MyProfileDTO {
        try await client.send(.delete, "api/profiles/me/photos/\(id)")
    }

    /// Reihenfolge der eigenen Fotos - photos[0] ist das Hauptfoto.
    func reorderPhotos(_ body: ReorderPhotosRequestDTO) async throws -> MyProfileDTO {
        try await client.send(.put, "api/profiles/me/photos/order", body: body)
    }

    func updateNotificationSettings(
        _ body: NotificationSettingsRequestDTO
    ) async throws -> MyProfileDTO {
        try await client.send(.patch, "api/profiles/me/notifications", body: body)
    }

    // MARK: - notifications.py

    /// Bereitliegende App-Benachrichtigungen abholen.
    ///
    /// `X-Flexr-Background` weist den Abruf als Hintergrundabgleich aus: ohne
    /// den Header wuerde ausgerechnet der Lauf, der die Inaktivitaets-Erinnerung
    /// ausliefern soll, den Nutzer dauerhaft als aktiv erscheinen lassen.
    func pendingNotifications() async throws -> [PushNotificationDTO] {
        try await client.send(
            .get, "api/notifications/pending",
            headers: ["X-Flexr-Background": "1"]
        )
    }

    /// Gerätetoken für echte Push-Zustellung anmelden.
    ///
    /// Bei jedem Start, nicht nur nach dem Anmelden: iOS vergibt den Token neu,
    /// wenn die App neu installiert oder wiederhergestellt wird. Mehrfaches
    /// Anmelden desselben Tokens ist deshalb der Normalfall und schreibt
    /// serverseitig nur dieselbe Zeile fort.
    func registerPushToken(platform: String, token: String) async throws {
        try await client.send(
            .post,
            "api/notifications/token",
            body: PushTokenRequestDTO(platform: platform, token: token)
        )
    }

    func unregisterPushToken(platform: String, token: String) async throws {
        try await client.send(
            .delete,
            "api/notifications/token",
            body: PushTokenRequestDTO(platform: platform, token: token)
        )
    }

    func markNotificationsDelivered(_ body: MarkDeliveredRequestDTO) async throws {
        try await client.send(
            .post, "api/notifications/delivered",
            body: body,
            headers: ["X-Flexr-Background": "1"]
        )
    }

    // MARK: - swipes.py

    func deck() async throws -> [ProfileDTO] {
        try await client.send(.get, "api/swipes/deck")
    }

    func swipe(_ body: SwipeRequestDTO) async throws -> SwipeResultDTO {
        try await client.send(.post, "api/swipes", body: body)
    }

    /// Wer mich geliket hat. Ohne Premium kommt keine Fehlermeldung, sondern
    /// nur die Anzahl ohne Profile (`premium_required = true`).
    func incomingLikes() async throws -> IncomingLikesDTO {
        try await client.send(.get, "api/swipes/incoming")
    }

    /// Letzten Swipe zurücknehmen — nur mit Premium, sonst 403.
    func rewindLastSwipe() async throws -> RewindResultDTO {
        try await client.send(.post, "api/swipes/rewind")
    }

    // MARK: - matches.py / messages.py

    func matches() async throws -> [MatchDTO] {
        try await client.send(.get, "api/matches")
    }

    func unmatch(matchID: String) async throws {
        try await client.send(.delete, "api/matches/\(matchID)")
    }

    func messages(matchID: String) async throws -> [MessageDTO] {
        try await client.send(.get, "api/matches/\(matchID)/messages")
    }

    func sendMessage(matchID: String, body: SendMessageRequestDTO) async throws -> MessageDTO {
        try await client.send(.post, "api/matches/\(matchID)/messages", body: body)
    }

    func clearMessages(matchID: String) async throws {
        try await client.send(.delete, "api/matches/\(matchID)/messages")
    }

    /// „Chat löschen": Match, Swipe und Nachrichten bleiben bestehen — nur die
    /// Unterhaltung verschwindet aus der Chats-Übersicht (siehe `in_chats`).
    func deleteChat(matchID: String) async throws {
        try await client.send(.delete, "api/matches/\(matchID)/chat")
    }

    // MARK: - billing.py

    func membershipStatus() async throws -> MembershipStatusDTO {
        try await client.send(.get, "api/billing/status")
    }

    func createCheckout(_ body: CheckoutRequestDTO) async throws -> CheckoutUrlDTO {
        try await client.send(.post, "api/billing/checkout", body: body)
    }

    func createPortal() async throws -> PortalUrlDTO {
        try await client.send(.post, "api/billing/portal")
    }

    /// Eine signierte StoreKit-Transaktion einreichen.
    ///
    /// Wird nach jedem Kauf aufgerufen und außerdem beim Start für alle
    /// laufenden Berechtigungen. Mehrfaches Einreichen desselben Belegs ist
    /// ausdrücklich vorgesehen und schreibt serverseitig nur dieselbe Zeile
    /// fort — daran hängt die Wiederherstellung nach einem Gerätewechsel.
    func submitAppleTransaction(_ signedTransaction: String) async throws -> StorePurchaseResultDTO {
        try await client.send(
            .post,
            "api/billing/apple/transaction",
            body: AppleTransactionRequestDTO(signedTransaction: signedTransaction)
        )
    }

    // MARK: - safety.py

    func report(_ body: ReportRequestDTO) async throws -> ReportAckDTO {
        try await client.send(.post, "api/reports", body: body)
    }

    func moderationNotice() async throws -> ModerationNoticeDTO? {
        try await client.sendOptional(.get, "api/moderation/notice", as: ModerationNoticeDTO.self)
    }

    func block(_ body: BlockRequestDTO) async throws {
        try await client.send(.post, "api/blocks", body: body)
    }

    func listBlocks() async throws -> [String] {
        try await client.send(.get, "api/blocks")
    }

    /// Wie `listBlocks()`, aber mit Name, Alter und Vorschaubild statt nur der
    /// IDs. Eigene Methode statt eines Parameters an `listBlocks()`, damit der
    /// Server-Standard (ohne `?detail=true`) unverändert die reine ID-Liste
    /// bleibt — siehe Docstring in `backend/app/routers/safety.py`.
    func listBlockedUsers() async throws -> [BlockedUserOutDTO] {
        try await client.send(.get, "api/blocks", query: ["detail": "true"])
    }

    func unblock(userID: String) async throws {
        try await client.send(.delete, "api/blocks/\(userID)")
    }

    // MARK: - withdrawal.py

    /// Online-Rücktrittsfunktion (§ 13a FAGG) — direkt aus der App statt aus
    /// dem Formular auf flexr.social/widerruf.html. Braucht keine Anmeldung;
    /// ein mitgeschickter Token ordnet die Erklärung nur zusätzlich dem Konto zu.
    func declareWithdrawal(_ body: WithdrawalRequestDTO) async throws -> WithdrawalAckDTO {
        try await client.send(.post, "api/withdrawal", body: body)
    }

    // MARK: - gyms.py

    func searchGyms(query: String) async throws -> [GymDTO] {
        try await client.send(.get, "api/gyms", query: ["q": query])
    }

    func suggestGym(_ body: GymSuggestRequestDTO) async throws -> GymDTO {
        try await client.send(.post, "api/gyms/suggest", body: body)
    }

    // MARK: - verification.py

    func verificationStatus() async throws -> VerificationStatusDTO {
        try await client.send(.get, "api/verification/status")
    }

    func startVerification() async throws -> VerificationStatusDTO {
        try await client.send(.post, "api/verification/start")
    }

    func presignSelfie(_ body: PresignPhotoRequestDTO) async throws -> PresignPhotoResponseDTO {
        try await client.send(.post, "api/verification/selfies/presign", body: body)
    }

    func submitVerification(_ body: VerificationSubmitRequestDTO) async throws -> VerificationStatusDTO {
        try await client.send(.post, "api/verification/submit", body: body)
    }

    /// Schritt 2: amtlicher Lichtbildausweis, privat abgelegt und nur temporär.
    func presignDocument(
        _ body: VerificationDocumentPresignRequestDTO
    ) async throws -> PresignPhotoResponseDTO {
        try await client.send(.post, "api/verification/document/presign", body: body)
    }

    func submitDocument(
        _ body: VerificationDocumentSubmitRequestDTO
    ) async throws -> VerificationStatusDTO {
        try await client.send(.post, "api/verification/document/submit", body: body)
    }

    /// Eingereichte Aufnahmen zurückziehen, solange niemand geprüft hat.
    func discardDocuments() async throws -> VerificationStatusDTO {
        try await client.send(.delete, "api/verification/document")
    }

    // MARK: - E-Mail-Bestätigung (email_verify.py)

    /// Neuen Aktivierungslink anfordern (nur für unbestätigte Adressen).
    func resendVerificationEmail() async throws -> EmailResendResponseDTO {
        try await client.send(.post, "api/auth/email/resend")
    }

    /// Token aus dem Aktivierungslink einlösen — braucht keine Anmeldung.
    func confirmEmail(_ body: EmailConfirmRequestDTO) async throws -> EmailConfirmResponseDTO {
        try await client.send(.post, "api/auth/email/confirm", body: body)
    }

    // MARK: - Objekt-Storage (Presigned PUT, absolute URL)

    /// Lädt eine Bilddatei direkt in den Objekt-Storage. Es fließen keine
    /// Bilddaten durchs Backend.
    func upload(to presignedURL: String, contentType: String, data: Data) async throws {
        try await client.upload(to: presignedURL, contentType: contentType, data: data)
    }
}

/// Ortsermittlung zu einer Postleitzahl.
///
/// Als Protokoll, damit [PlzRepository] im Test ohne Netz auskommt — die
/// Entsprechung des Retrofit-Interfaces in der Android-App.
protocol PostalCodeLookup {
    func lookupPostalCode(_ plz: String) async throws -> PlzLookupDTO
}

/// PLZ-Lookup über das eigene Backend (GET /api/geo/plz/{plz}).
///
/// Früher fragten alle Clients openplzapi.org direkt an. Das war weder
/// verlässlich (der Dienst weist manche HTTP-Clients mit 418 ab) noch korrekt
/// (die seitenweise Ortschaftsliste ließ die Heuristik bei großen PLZ
/// danebengreifen). Das Backend liefert den amtlichen Ortsnamen.
struct BackendPlzAPI: PostalCodeLookup {

    private let client: APIClient

    init(client: APIClient) {
        self.client = client
    }

    func lookupPostalCode(_ plz: String) async throws -> PlzLookupDTO {
        try await client.send(.get, "api/geo/plz/\(plz)")
    }
}

File: android-native/app/src/main/java/flexr/social/app/data/remote/FlexrApi.kt
package flexr.social.app.data.remote

import flexr.social.app.data.remote.dto.AddPhotoRequestDto
import flexr.social.app.data.remote.dto.BlockRequestDto
import flexr.social.app.data.remote.dto.CheckoutUrlDto
import flexr.social.app.data.remote.dto.DeleteAccountRequestDto
import flexr.social.app.data.remote.dto.GymDto
import flexr.social.app.data.remote.dto.GymSuggestRequestDto
import flexr.social.app.data.remote.dto.LoginRequestDto
import flexr.social.app.data.remote.dto.MatchDto
import flexr.social.app.data.remote.dto.MembershipStatusDto
import flexr.social.app.data.remote.dto.MessageDto
import flexr.social.app.data.remote.dto.ModerationNoticeDto
import flexr.social.app.data.remote.dto.MyProfileDto
import flexr.social.app.data.remote.dto.PlzLookupDto
import flexr.social.app.data.remote.dto.PortalUrlDto
import flexr.social.app.data.remote.dto.PresignPhotoRequestDto
import flexr.social.app.data.remote.dto.PresignPhotoResponseDto
import flexr.social.app.data.remote.dto.ProfileDto
import flexr.social.app.data.remote.dto.RegisterRequestDto
import flexr.social.app.data.remote.dto.ReportAckDto
import flexr.social.app.data.remote.dto.ReportRequestDto
import flexr.social.app.data.remote.dto.SendMessageRequestDto
import flexr.social.app.data.remote.dto.SwipeRequestDto
import flexr.social.app.data.remote.dto.SwipeResultDto
import flexr.social.app.data.remote.dto.TokenResponseDto
import flexr.social.app.data.remote.dto.VerificationStatusDto
import flexr.social.app.data.remote.dto.VerificationSubmitRequestDto
import okhttp3.RequestBody
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.HTTP
import retrofit2.http.Header
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query
import retrofit2.http.Url

/**
 * Vollständige Abbildung der FastAPI-Router aus backend/app/routers/.
 * Ein Endpunkt pro Methode, gleiche Pfade wie im Web-Frontend.
 */
interface FlexrApi {

    // ---------- auth.py ----------

    @POST("api/auth/register")
    suspend fun register(@Body body: RegisterRequestDto): TokenResponseDto

    @POST("api/auth/login")
    suspend fun login(@Body body: LoginRequestDto): TokenResponseDto

    // ---------- profiles.py ----------

    @GET("api/profiles/me")
    suspend fun getMyProfile(): MyProfileDto

    @PATCH("api/profiles/me")
    suspend fun updateMyProfile(@Body body: flexr.social.app.data.remote.dto.UpdateProfileRequestDto): MyProfileDto

    /** DELETE mit Body — Retrofits @DELETE erlaubt das nicht, @HTTP schon. */
    @HTTP(method = "DELETE", path = "api/profiles/me", hasBody = true)
    suspend fun deleteMyAccount(@Body body: DeleteAccountRequestDto)



    @POST("api/profiles/me/photos/presign")
    suspend fun presignPhoto(@Body body: PresignPhotoRequestDto): PresignPhotoResponseDto

    @POST("api/profiles/me/photos")
    suspend fun addPhoto(@Body body: AddPhotoRequestDto): MyProfileDto

    @DELETE("api/profiles/me/photos/{photoId}")
    suspend fun deletePhoto(@Path("photoId") photoId: String): MyProfileDto

    // ---------- swipes.py ----------

    @GET("api/swipes/deck")
    suspend fun getDeck(): List<ProfileDto>

    @POST("api/swipes")
    suspend fun swipe(@Body body: SwipeRequestDto): SwipeResultDto

    // ---------- matches.py / messages.py ----------

    @GET("api/matches")
    suspend fun getMatches(): List<MatchDto>

    @DELETE("api/matches/{matchId}")
    suspend fun unmatch(@Path("matchId") matchId: String)

    @GET("api/matches/{matchId}/messages")
    suspend fun getMessages(@Path("matchId") matchId: String): List<MessageDto>

    @POST("api/matches/{matchId}/messages")
    suspend fun sendMessage(
        @Path("matchId") matchId: String,
        @Body body: SendMessageRequestDto,
    ): MessageDto

    @DELETE("api/matches/{matchId}/messages")
    suspend fun clearMessages(@Path("matchId") matchId: String)

    // ---------- billing.py ----------

    @GET("api/billing/status")
    suspend fun getMembershipStatus(): MembershipStatusDto

    @POST("api/billing/checkout")
    suspend fun createCheckout(): CheckoutUrlDto

    @POST("api/billing/portal")
    suspend fun createPortal(): PortalUrlDto

    // ---------- safety.py ----------

    @POST("api/reports")
    suspend fun report(@Body body: ReportRequestDto): ReportAckDto

    @GET("api/moderation/notice")
    suspend fun moderationNotice(): ModerationNoticeDto?

    @POST("api/blocks")
    suspend fun block(@Body body: BlockRequestDto)

    @GET("api/blocks")
    suspend fun listBlocks(): List<String>

    @DELETE("api/blocks/{userId}")
    suspend fun unblock(@Path("userId") userId: String)

    // ---------- gyms.py ----------

    @GET("api/gyms")
    suspend fun searchGyms(@Query("q") query: String): List<GymDto>

    @POST("api/gyms/suggest")
    suspend fun suggestGym(@Body body: GymSuggestRequestDto): GymDto

    // ---------- geo.py ----------

    @GET("api/geo/plz/{plz}")
    suspend fun lookupPostalCode(@Path("plz") plz: String): PlzLookupDto

    // ---------- verification.py ----------

    @GET("api/verification/status")
    suspend fun getVerificationStatus(): VerificationStatusDto

    @POST("api/verification/start")
    suspend fun startVerification(): VerificationStatusDto

    @POST("api/verification/selfies/presign")
    suspend fun presignSelfie(@Body body: PresignPhotoRequestDto): PresignPhotoResponseDto

    @POST("api/verification/submit")
    suspend fun submitVerification(@Body body: VerificationSubmitRequestDto): VerificationStatusDto

    // ---------- Objekt-Storage (Presigned PUT, absolute URL) ----------

    /**
     * Lädt eine Bilddatei direkt in den Objekt-Storage (S3/Cloudflare R2).
     * Bewusst absolut adressiert (@Url) — es fließen keine Bilddaten durchs Backend.
     */
    @PUT
    suspend fun uploadToPresignedUrl(
        @Url url: String,
        @Header("Content-Type") contentType: String,
        @Body body: RequestBody,
    )
}
